### ISSUE TITLE
Staunch mode should not stop for independant `&%>` failure

### ISSUE THREAD

**Issue by guibou** [ndmitchell/shake#512]
In the following simple example, I can build any `*.foo` file from an `*.md5` file, and any `*.md5` file from a `*` file. In this example, the rule to build `*.md5` from `*` files build them all at once.

I suppose I have an `a` and `b` file on the disk:

```haskell
#!/usr/bin/env stack
-- stack runghc --resolver lts-7.15 --package base --package shake --install-ghc -- -hide-all-packages
import Development.Shake
import Development.Shake.FilePath

md5Cmd :: FilePath -> Action ()
md5Cmd output = do
  let input = dropExtension output
  need [input]
  (Stdout outputString) <- command [] "md5sum" [input]
  writeFileChanged output outputString

main = shakeArgs (shakeOptions {shakeStaunch = True}) $ do
  want ["a.md5.foo", "b.md5.foo"]

  "*.md5.foo" %> \foo -> do
    let md5 = dropExtension foo
    need [md5]

    command_ [] "cp" [md5, foo]

  -- build the md5 files, however they are all build at once
  ["a.md5", "b.md5"] &%> \[a, b] -> do
    md5Cmd a

    -- Ho, no, the build is broken and cannot build *b*
    -- md5Cmd b
```

For the example, this build is deliberately broken and does not build `b.md5`. Staunch mode should be able to write `a.md5.foo`, because `a.md5` was built and it exits a rule to build `a.md5.foo` from `a.md5`, however it appears that the failure to build `b.md5` prevents the rule which is supposed to build `a.md5.foo` to run.

Thank you.

**Comment by ndmitchell** [comment]
The line:

    ["a.md5", "b.md5"] &%> \[a, b] -> do

Builds `a` and `b` simultaneously, saying that there is one action to build both. What it looks like you want is:

    ["a.md5", "b.md5"] |%> \a -> do
        md5Cmd a

Which says that there is one rule that is used for these two files, independently. That means if there is a failure the other can run in staunch mode, and if there is not they can run in parallel. However, you may find a more generic rule works better:

    "*.md5" %> \a -> do
        md5Cmd a

**Comment by guibou** [comment]
The example I gave is a simplified version of a more complex build system.

I have a command `oneCommandWhichDoEverything` which generates many output files at once. I cannot change that. Actually it takes an input file `input.project` and generates many files `a.exr`, `b.exr`, `c.exr`... My shake rules look like:

```haskell
["a.exr", "b.exr", "c.exr", "d.exr", "a.txt", "b.txt", "c.txt", "d.txt"] &%> \_ -> do
     need ["input.project"]
     oneCommandWhichDoEverything "input.project"

"*.png" %> \png -> do
     exrToPng (png -<.> "exr") png

want ["a.png", "b.png", "c.png", "d.png", "a.txt", "b.txt", "c.txt", "d.txt"]
```

The idea is that even if `c.exr` is not generated by the first rule, if `a.exr` is generated, I want `a.png` to be generated.

**Comment by ndmitchell** [comment]
As soon as `oneCommandWhichDoEverything` fails due to an exception, as far as Shake can tell, nothing was generated. If it so happens that `a.exr` is generated, Shake can't tell that, and even if it did see it update, it wouldn't know if it was the final update or not. Without breaking that command apart, it's hard for Shake to see inside that a subset succeeds. From `oneCommandWhichDoEverything`, do you yourself know what succeeded?

**Comment by guibou** [comment]
I can check that the files are available on disk (I'm deleting them prior to run `oneCommandWhichDoEverything`).

Actually I realize that I'm trying to use shake in a way for which it is not designed for. Perhaps I'll have more success by rewriting a specific tool for that purpose. I will just miss the shake report.

**Comment by ndmitchell** [comment]
It does indeed seem like you are abusing Shake here, but I think it can take a little bit more abuse before failing! You can probably write:

```
"output.project" %> \_ -> do
    need ["input.project"]
    oneCommandWhichDoEverything "input.project"
    writeFile' "output.project" ""
 
["a.exr", "b.exr", "c.exr", "d.exr", "a.txt", "b.txt", "c.txt", "d.txt"] |%> \_ -> do
     need ["output.project"]
```

Now make `oneCommandWhichDoEverything` always succeed, and the individual `a.txt` things will succeed or fail individually complaining they didn't get created or not. That way staunch mode should still work properly.

**Comment by guibou** [comment]
Actually I was using something like your last proposal, however this comes with another drawback (one last proof of my shake abuse), it is that if someone deletes "a.exr" without deleting "output.project", shake will never rebuild "output.project".

(I'm closing this issue, I guess it is "won't fix")

**Comment by ndmitchell** [comment]
Yep, that is a drawback - as soon as you start lying about dependencies things do start to go wrong...
